### ISSUE TITLE
Lets start support rpm based OS 

### ISSUE THREAD

**Issue by Joni-Lover** [aptly-dev/aptly#39]
many users would like to use your tool for RPM repositories

**Comment by dmitrievav** [comment]
+1

**Comment by kisoku** [comment]
I would also really love to see this make it into aptly. We have a bunch of ruby stuff wrapping aptly and createrepo right now that could go away if you had this functionality

**Comment by icebrian** [comment]
FYI Pulp provides repo management for RHEL/CentOS (http://www.pulpproject.org/) which is part of Katello (http://www.katello.org/) which provides content and lifecycle management and which is included in The Foreman (http://theforeman.org/).

May-be it would be interesting that aptly and Pulp/Katello colaborate? (Just a thought)

**Comment by elliotmoore** [comment]
-1, sorry.
I'd prefer built-in rpm support in aptly, rather than depend on 3 different projects each with their own dependancies..

**Comment by sepulworld** [comment]
@smira do you have a rough outline of what needs to be done to support RPMs? I would like to help out with this but need a little help getting started.

**Comment by smira** [comment]
@sepulworld I tried to collect some work items here: https://docs.google.com/document/d/1Ili6lDOJO3Epj__JTdVGiJKpLBrU-R0mckhdO5prKvk/edit?usp=sharing

It's not complete list

**Comment by redrampage** [comment]
+1

**Comment by 6opo9a** [comment]
+1

**Comment by Conan-Kudo** [comment]
@smira 

The [createrepo_c](https://github.com/Tojaj/createrepo_c)/[createrepo](http://createrepo.baseurl.org/) suite does quite a bit:

* Creating repositories
* Merging repositories
* Modifying them
* Generating DeltaRPMs and creating associated repository data for them

As far as I know, some of this functionality has never really existed in Debian systems, so if Aptly is too Debian specific, it may take quite a bit more work to implement the needed functionality to support RPM Metadata (rpm-md) repositories.

As for the createrepo suite, createrepo itself hasn't seen any development in a while, as efforts are largely focused on getting moved over to createrepo_c.

**Comment by riteshnanda09** [comment]
Hi ,

Do we have any update on this . When this feature is getting releases

**Comment by pquerna** [comment]
https://github.com/cavaliercoder/go-rpm provides native go code for dealing with RPM file format and parts of YUM which would be useful for this project.

**Comment by cavaliercoder** [comment]
Thanks @pquerna for the mention. My main goal for go-rpm was a yum/rpm implementation to backend [y10k](https://github.com/cavaliercoder/y10k). Having a look at aptly, my efforts with y10k seem to have a very similar use case, except targeting rpm instead of deb and obviously with much less maturity.

Currently (though unreleased), go-rpm and y10k (in the `native_rpm` branch) can mimic `reposync` in 100% native Go. I am currently reimplementing `createrepo` functionality natively with gradual success.

I'd be delighted to have support in developing go-rpm and y10k. I would also consider abandoning y10k and instead contributing to the development of aptly as an rpm compatible tool. My goal is better tooling for publishing yum repos.

**Comment by Conan-Kudo** [comment]
@cavaliercoder Why not instead use libcreaterepo_c with your program?

**Comment by cavaliercoder** [comment]
Fair question and still a good suggestion but like @elliotmoore, I prefer the idea of Go native implementations that have no dependencies on system libraries.

It's definitely a trade off between the stability of the original library (libcreaterepo_c) and having a Go native implementation that can be reused in different tools without creating platform limitations due to dependencies. That, and it's a pretty cool challenge and creates choice for developers.

**Comment by techzilla** [comment]
Aptly is leaps and bounds better written and designed than Pulp, it gets RPM support, company switches tomorrow.

**Comment by thebwt** [comment]
I think the problem is the whole concept within yum repos is that you can do whatever you want, where as debian repo's have a much more rigid hierarchy imposed on itself.

If this ran with the idea of imposing dist/arch/component paradigms on yum repos, would that blow people's mind too much?

it can be done, createrepo_c don't care.

**Comment by FlorianHeigl** [comment]
@thebwt 
It might be just be OK! At least for people who _primarily_ use dpkg/apt, since it'll be natural for them.
I've _never_ felt well in the debian structure of repos, but I think for people who like it should be good! I know enough people who feel more at home in how debian does them. Having the same structure across different OS is quite helpful.


(I'm in testing of repositor, which is kind of a clone of Pulp v1, meaning just the pure repo management. That one looks like pulp and just is more manageable. Think about if you want to be compatible with people who rather want the unstructured yum way. If not, maybe it's not even wrong to have two tools for two mindsets. Well, three since there's also people who enjoy all the spaceshipness of Pulp)

**Comment by techzilla** [comment]
I would be very happy with increased structure, my only concern would be on mirroring interop, but for hosted repo's I already separate them out Debian style.   Many yum repos themselves have a structure to them, implicitly, you could simply mirror a specified distro/arch specific portion of a repo.

**Comment by james-lawrence** [comment]
seems like this has died? is anyone actively working on it?

**Comment by FlorianHeigl** [comment]
@james-lawrence there was doubts if it's really in scope for aptly to support a very different repo structure. I think the decision is still open, and your input (needs) would help. If you need movement on this, please write them down.

@techzilla I fully understand you. I think Pulp v1 was really good, and it seems the new team never really read the original paper and instead of a tool build a star destroyer... 
In one place we use Pulp and will keep using it because we simply need the scalability it has, and especially the faster repo updates it does (better than createrepo -c -u or so). but for normal sized use cases it's too bad at supporting the actual repo workflow (vs. supporting putting 2 fulltime devs on implementing+maintaining the workflow using the API), and comes with just too sick overheads.

I did also try repositor.io by now. It is a fragile clusterfuck and would need more maintainers to keep up it's promises (docker registry? yeah, but not if you need to have an authentication or would like to work with a recent client? errata? yeah, but the daily updates last happened in 2016, ...)

So how do I put this - there is still room for doing things better.
But I would recommend to not underestimate the effort of doing twice as much, and still having a unified interface to manage very different concepts of repos.

**Comment by james-lawrence** [comment]
@FlorianHeigl not really worried about the structure imposed. just looking for a solid tool that:

1) easy to deploy
2) is usable using standard yum/dnf .repo configuration files.
3) supports uploading packages/mirror other rpm repos

**Comment by Conan-Kudo** [comment]
It would also be nice if it had the ability to generate a detached signature of the repomd.xml for distributions that generally expect it (openSUSE, etc.).

**Comment by FlorianHeigl** [comment]
@Conan-Kudo good point - we manually do that in Pulp via something cron-like and its not nice.
from experience: you'd really want that well-integrated so that you don't end up with invalid signatures if keys are not loaded.
Ideally, aptly would be able to read/look at the signatures, anything "after the fact" is highly fragile.

**Comment by lazyfrosch** [comment]
We actually use aptly to upload RPM packages.

The packages are then published to a RPM repository via a local cron job that picks up the upload.

**Comment by maxadamo** [comment]
createrepo does not support snapshot

**Comment by szhem** [comment]
Hi there, as the landing page states that support of yum repositories coming soon, I'm just wondering where there is any ETA regarding this feature?

![image](https://user-images.githubusercontent.com/1523889/37905689-5499cc40-3108-11e8-841c-33ddcd43a999.png)

**Comment by carlwgeorge** [comment]
@smira If that's the case, please remove mention of RPM support from the website.

**Comment by Conan-Kudo** [comment]
@smira If you want to implement RPM repositories in a "pure-Go" manner, you can look at the following to do RPM parsing:
* https://github.com/sassoftware/go-rpmutils
* https://github.com/cavaliercoder/go-rpm

The Yum repodata format is just XML, and there is plenty of test data in the `createrepo_c` repo: https://github.com/rpm-software-management/createrepo_c

**Comment by smira** [comment]
@Conan-Kudo yes, generating (publishing) repository should be fairly straightforward, but getting all the details right for rpm packages, making sure mirroring work, all the .deb/.rpm separation works properly is a lot of work...

**Comment by cavaliercoder** [comment]
Implementing the YUM/repo side of things in native Go is actually pretty heavy work. I made a solid start in https://github.com/cavaliercoder/go-rpm but the cost outweighed the benefit. A big part of the problem was that (at the time) there was no Go native implementations of SQLite or XZ compression. Things may have improved now. It's probably not worth the effort vs. calling out to createrepo.

**Comment by cavaliercoder** [comment]
Incomplete YUM implementation is hidden here: https://github.com/cavaliercoder/y10k/tree/native_rpm/yum

**Comment by Conan-Kudo** [comment]
@cavaliercoder You don't need SQLite metadata. DNF and Zypper only read the XML metadata and the XML metadata can be compressed in gzip, bzip2, or xz format currently. Yum will use XML metadata by default, too.

**Comment by cavaliercoder** [comment]
Cool! One of my goals was universal client support, so supporting most of the database formats was necessary. If newer clients only use the XML db, maybe that issue is de-scoped.

My projects are mostly unmaintained now - certainly not under active development. I'd just like to share what I found and hopefully someone smarter than me can bring things to a more valuable conclusion.

**Comment by Conan-Kudo** [comment]
@cavaliercoder All versions of DNF, Zypper, Yum, Smart, and APT-RPM use XML metadata. The SQLite metadata was an optimization added for external infrastructure tools to use (e.g. `repoview`, et al).

**Comment by cavaliercoder** [comment]
Great! Hopefully the way forward is simpler than I anticipate.

**Comment by bby-bishopclark** [comment]
> Great! Hopefully the way forward is simpler than I anticipate.

yeah.  Check out the apt-rpm work (apt4rpm on sourceforget) that Conectiva did 20 years ago.  It was fantastic.

**Comment by j1mc** [comment]
FWIW, I found this:  https://github.com/zbindenren/gym

I have not tested it, but it's at least more recent than the current progress within aptly, and it has several releases.

I don't see a LICENSE file included in the repo, but it would likely be worth reaching-out to the developer if anyone were interested in this.

**Comment by FlorianHeigl** [comment]
@j1mc the Repo itself has a MIT license set, so that one's easy.
during the 8-ish years that I followed this issue I had also tried out [repositorio](https://github.com/RexOps/repositorio) which has its limitations but is an OK choice for heterogenous environments. There's support for apt/yum/zypper.
It'll not be as efficient as i.e. Pulp if you want to freeze / clone many distro sized repos, but it makes more sense than waiting for aptly to be bent into something that it isn't.
and if one has a homogenous apt env, then the aptly choice is clear anyway.
